File: solutions/fsharp/sieve/Sieve.fs
module Sieve

let primes limit =
    let isPrime n =
        if n < 2 then false
        else 
            let root = int (sqrt (float n))
            seq {2..root} |> Seq.exists (fun x -> n % x = 0) |> not
    seq {2..limit} |> Seq.filter isPrime |> Seq.toList

File: solutions/scala/sieve/src/main/scala/Sieve.scala
object Sieve:
  def primes(limit: Int): List[Int] =
    val unmarkedNumbers = scala.collection.mutable.BitSet.fromSpecific(2 to limit)
    val limitSqrt = scala.math.sqrt(limit).toInt
    
    for
      candidate <- 2 to limitSqrt
      if unmarkedNumbers.contains(candidate)
      multiple <- candidate * candidate to limit by candidate
    do
      unmarkedNumbers -= multiple

    unmarkedNumbers.toList      
